Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs

```diff
@@ -290,35 +290,34 @@ public bool TryAddSolution(EntityUid targetUid, Solution? targetSolution, Soluti
         return solutionsMgr.Solutions[name];
     }
 
-    public string[] RemoveEachReagent(EntityUid uid, Solution solution, FixedPoint2 quantity)
+    /// <summary>
+    ///     Removes an amount from all reagents in a solution, adding it to a new solution.
+    /// </summary>
+    /// <param name="uid">The entity containing the solution.</param>
+    /// <param name="solution">The solution to remove reagents from.</param>
+    /// <param name="quantity">The amount to remove from every reagent in the solution.</param>
+    /// <returns>A new solution containing every removed reagent from the original solution.</returns>
+    public Solution RemoveEachReagent(EntityUid uid, Solution solution, FixedPoint2 quantity)
     {
-        var removedReagent = new string[solution.Contents.Count];
         if (quantity <= 0)
-            return Array.Empty<string>();
+            return new Solution();
 
-        var pos = 0;
-        for (var i = 0; i < solution.Contents.Count; i++)
+        var removedSolution = new Solution();
+
+        // We have to make a clone of contents as the original list might change...
+        var contents = solution.Contents.ToArray();
+        for (var i = 0; i < contents.Length; i++)
```

Comment:
why not just
```suggestion
        foreach (var (reagentId, _) in contents)
```

Reply:
Isn't the length liable to change via RemoveReagent?

Reply:
That was the whole reason for the `contents = solution.Contents.ToArray();`
Though alternatively, wouldn't this also just be solved by doing a for loop that starts from the end of the array?

Reply:
> why not just [foreach]

Ye, that's better. I was refactoring the original and didn't even think about turning it into a `foreach`, oops.

> Though alternatively, wouldn't this also just be solved by doing a for loop that starts indexing from the end of `solution.Contents`, so that the ToArray() wouldn't be needed?

That would be the best option, so I'll do that

Reply:
Done now.